src/logic/player.h

```diff
@@ -692,6 +700,8 @@ class Player {
 
 	PlayerBuildingStats building_stats_;
 
+	std::map<std::tuple<unsigned, unsigned, unsigned, unsigned>, uint32_t> soldier_stats_;
```

Comment:
This is fragile because all tuple members have the same data type. How about introducing a new struct for a Soldier's attributes, like this:
```
struct SoldierLevels {
    unsigned health;
    unsigned attack;
    unsigned defense;
    unsigned evade;

    unsigned total() const;   
}
```
This could then be used both in the Solder class to return the Solder's attributes, and here.